atomic_reactor/plugins/post_compare_components.py

```diff
@@ -0,0 +1,106 @@
+from atomic_reactor.constants import (PLUGIN_COMPARE_COMPONENTS_KEY,
+                                      PLUGIN_FETCH_WORKER_METADATA_KEY)
+
+
+supported_types = ["rpm"]
+
+
+class CompareComponentsPlugin(PostBuildPlugin):
+    """
+    Compare components from each worker build and verify the same version was
+    on each worker.
+    """
+
+    key = PLUGIN_COMPARE_COMPONENTS_KEY
+    is_allowed_to_fail = False
+
+    def rpm_compare(self, a, b):
+        """
+        Compare rpm component version 'a' with 'b'.
+        'name' is implied equal as it is the key() lookup
+        """
+        if a['version'] == b['version'] and \
+           a['release'] == b['release'] and \
+           a['signature'] == b['signature']:
+            return
+
+        raise ValueError("%s != %s" % (a, b))
+
+    def get_component_list(self, worker_metadatas):
+        """
+        Find the worker build's component list.
+        The strategy is to find the 'output' key in the metadata and then
+        search the array for a 'docker-image' type.  The 'log' types are not
+        interesting.
+
+        :return: array of components
+        """
+        comp_list = []
+        for platform in sorted(worker_metadatas.keys()):
+            for instance in worker_metadatas[platform]['output']:
+                if instance['type'] == 'docker-image':
+                    if 'components' not in instance or not instance['components']:
+                        self.log.warn("Missing 'components' key in 'output' metadata instance: %s",
+                                      instance)
+                        continue
+
+                    comp_list.append(instance['components'])
+
+        return comp_list
+
+    def run(self):
+        """
+        Run the plugin.
+        """
+
+        worker_metadatas = self.workflow.postbuild_results.get(PLUGIN_FETCH_WORKER_METADATA_KEY)
+        comp_list = self.get_component_list(worker_metadatas)
+
+        if not comp_list:
+            raise ValueError("No components to compare")
+
+        # master compare list
+        master_comp = {}
+
+        # The basic strategy is to start with empty lists and add new component
+        # versions as we find them.  Upon next iteration, we should notice
+        # duplicates and be able to compare them.  If the match fails, we raise
+        # an exception.  If the component name does not exist, assume it was an
+        # arch dependency, add it to list and continue.  By the time we get to
+        # the last arch, we should have every possible component in the master
+        # list to compare with.
+
+        # Keep everything separated by component type
+        failed = False
+        for components in comp_list:
+            for component in components:
+                t = component['type']
+                name = component['name']
+
+                if t not in supported_types:
+                    raise ValueError("Type %s not supported")
```

Comment:
I guess this is fine. It means we'll need to add support for non-RPM content to this plugin at the same time as adding metadata tracking to fetch_maven_artifacts. @lcarva, OK with that?

Reply:
I guess that 's fine.

Reply:
I am fine with either an exception or warning.  I have to go into the code to change it to immutable anyway..

Reply:
Let's make it an exception then. Adding non-RPM content will be a code change in any case, so if we decide otherwise when we do that we could always change this to a warning at that time.